### PR TITLE
scalding-quotation refactorings

### PR DESCRIPTION
**Pull request by fwbrasil**
This PR addresses @ttim's feedback on https://github.com/twitter/scalding/pull/1755

### CODE REVIEW

File: scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala [hunk 1]
```diff
@@ -8,6 +8,71 @@ case class TypeName(asString: String) extends AnyVal
...
+
+    def isSubclass(c: TypeName) =
+      try
+        superClass.isAssignableFrom(Class.forName(c.asString))
+      catch {
+        case _: ClassNotFoundException =>
+          false
+      }
+
+    def loop(p: Projection): Either[Projection, Option[Projection]] =
```
Comment (ttim) [inline scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala]: I still find this logic complicated. But comment makes everything better in this case!

File: scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala [hunk 2]
```diff
@@ -30,81 +95,29 @@ final case class Property(path: Projection, accessor: Accessor, typeName: TypeNa
...
-      def isSubclass(c: TypeName) =
-        try
-          superClass.isAssignableFrom(Class.forName(c.asString))
-        catch {
-          case _: ClassNotFoundException =>
-            false
-        }
-
-      def loop(p: Projection): Either[Projection, Option[Projection]] =
+      @tailrec def rootProjection(p: Projection): TypeReference =
```
Comment (ttim) [inline scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala]: Move to `Projection` as well?

**Review by ttim — APPROVED** [review]
Thanks for the PR!

File: scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala (hunk 2 above)
Comment (fwbrasil) [reply scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala]: done

File: scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala [hunk 3]
```diff
@@ -8,6 +8,87 @@ case class TypeName(asString: String) extends AnyVal
...
     Property(this, accessor, typeName)
+
+  /**
+   * Used to check if a projection is based on a specific type.
+   * For instance, given the type name `Person`:
+   *
+   * `Person.name`   => true
+   * `Contact.phone` => false
+   */
+  def isBasedOntype(typeName: TypeName): Boolean = {
```
Comment (ttim) [inline scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala]: nit: `type` should start with capital letter.

Personally I prefer to see rootProjection method here and just inline `rootProjection(this).typeName == typeName` at callsite.

Comment (fwbrasil) [reply scalding-quotation/src/main/scala/com/twitter/scalding/quotation/Projection.scala]: done

**Comment by fwbrasil** [comment]
the CI failed with an unrelated error. Could you someone with permission restart the job? https://travis-ci.org/twitter/scalding/jobs/319461956

**Pull request merged**